src/rgw/services/svc_account_rados.cc

```diff
@@ -0,0 +1,274 @@
+                                            bool exclusive,
+                                            std::map<std::string, bufferlist> *pattrs,
+                                            optional_yield y)
+{
+  bufferlist data_bl;
+  encode(info, data_bl);
+
+  RGWSI_MBSObj_PutParams params(data_bl, pattrs, mtime, exclusive);
+  return svc.meta_be->put(_ctx, get_meta_key(info), params, objv_tracker, y, dpp);
+}
+
+int RGWSI_Account_RADOS::read_account_info(const DoutPrefixProvider* dpp,
+                                           RGWSI_MetaBackend::Context *ctx,
+                                           const std::string& account_id,
+                                           RGWAccountInfo *info,
+                                           RGWObjVersionTracker * const objv_tracker,
+                                           real_time * const pmtime,
+                                           std::map<std::string, bufferlist> * const pattrs,
+                                           optional_yield y)
+{
+  bufferlist bl;
+  RGWSI_MBSObj_GetParams params(&bl, pattrs, pmtime);
+  int r = svc.meta_be->get_entry(ctx, account_id, params, objv_tracker, y, dpp);
+  if (r < 0) {
+    return r;
+  }
+
+  auto bl_iter = bl.cbegin();
+  try {
+    decode(*info, bl_iter);
+    if (info->get_id() != account_id) {
+      lderr(svc.meta_be->ctx()) << "ERROR: read_account_info account id mismatch" << info->get_id() << "!= " << account_id << dendl;
+      return -EIO;
+    }
+  } catch (buffer::error& err) {
+    ldout(svc.meta_be->ctx(), 0) << "ERROR: failed to decode account info, caught buffer::error" << dendl;
+    return -EIO;
+  }
+
+  return 0;
+}
+
+int RGWSI_Account_RADOS::remove_account_info(const DoutPrefixProvider* dpp,
+                                             RGWSI_MetaBackend::Context *ctx,
+                                             const std::string& account_id,
+                                             RGWObjVersionTracker *objv_tracker,
+                                             optional_yield y)
+{
+  RGWSI_MBSObj_RemoveParams params;
+  int ret = svc.meta_be->remove(ctx, account_id, params, objv_tracker, y, dpp);
+  if (ret <0 && ret != -ENOENT && ret != -ECANCELED) {
+    ldpp_dout(dpp, 0) << "ERROR: could not remove account: " << account_id << dendl;
+    return ret;
+  }
+  return 0;
+}
+
+struct rgw_account_user_header {
+  uint32_t current_users {0};
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(current_users, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(current_users, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(rgw_account_user_header)
+
+
+int RGWSI_Account_RADOS::add_user(const DoutPrefixProvider* dpp,
+                                  const RGWAccountInfo& info,
+                                  const rgw_user& user,
+                                  optional_yield y)
```

Comment:
`add_user()` and `remove_user()` would be better implemented in cls to avoid the extra round trips for omap_get_header and omap_set_header, which aren't atomic in their increment/decrement of `current_users`

Reply:
are `add_user()` and `remove_user()` what the `rgw::sal::Account` `link_user()` and `unlink_user()` functions going to use under the hood?